In [3]:
import pandas as pd
import numpy as np
import seaborn  as sns
import matplotlib.pyplot as plt 

##EDA

In [21]:
red_wine = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv',sep=";")
white_wine = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv',sep=";")

Creates a new column called 'quality label' based on the score in the existing column 'quality'

In [41]:
red_wine['quality_label'] = red_wine['quality'].apply(lambda value: 'low'
if value <= 5 else 'medium'
if value <= 7 else 'high')

In [42]:
white_wine['quality_label'] = white_wine['quality'].apply(lambda value: 'low'
if value <= 5 else 'medium'
if value <= 7 else 'high')

Combines the red and white wine dfs

In [50]:
combined=pd.concat([red_wine, white_wine], ignore_index=True)
combined.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
6492            6.2              0.21         0.29             1.6      0.039   
6493            6.6              0.32         0.36             8.0      0.047   
6494            6.5              0.24         0.19             1.2      0.041   
6495            5.5              0.29         0.30             1.1      0.022   
6496            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
6492                 24.0                  92.0  0.99114  3.27       0.50   
6493                 57.0                 168.0  0.99490  3.15       0.46   
6494                 30.0                 111.0  0.99254  2.99       0.46   
6495                 20.0                 110.0  0.98869  3.34       0.38   
6496                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality quality_label  
6492     11.2        6        medium  
6493      9.6        5           low  
6494      9.4        6        medium  
6495     12.8        7        medium  
6496     11.8        6        medium

In [51]:
combined.drop_duplicates(inplace=True, ignore_index=True)

Encode the values from 'quality label', creates a new column with the mapped labels

In [52]:

label_mapping = {'low': 0, 'medium': 1, 'high': 2}

combined['encoded_quality_label'] = combined['quality_label'].map(label_mapping) 

print(combined[['quality_label', 'encoded_quality_label']])


     quality_label  encoded_quality_label
0              low                      0
1              low                      0
2              low                      0
3           medium                      1
4              low                      0
...            ...                    ...
5313        medium                      1
5314           low                      0
5315        medium                      1
5316        medium                      1
5317        medium                      1

[5318 rows x 2 columns]


##Encoding categorical variables-->Ordinal variable

In [45]:
#from sklearn.preprocessing import LabelEncoder--> NOT FOR CATHGORICAL CLASSES 

#combined_wine = pd.concat([red_wine, white_wine], ignore_index=True)

# Select the column to be encoded
#C = combined_wine[['quality_label']]

# Create and fit the encoder 
#label_encoder = LabelEncoder()
#encoded_labels = label_encoder.fit_transform(C['quality_label'])

# Add the encoded labels as a new column
#combined_wine['encoded_quality_label'] = encoded_labels

# Save the DataFrame with encoded column
#combined_wine.to_csv('combined_wine_encoded.csv', index=False)


In [53]:
combined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5318 entries, 0 to 5317
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   fixed acidity          5318 non-null   float64
 1   volatile acidity       5318 non-null   float64
 2   citric acid            5318 non-null   float64
 3   residual sugar         5318 non-null   float64
 4   chlorides              5318 non-null   float64
 5   free sulfur dioxide    5318 non-null   float64
 6   total sulfur dioxide   5318 non-null   float64
 7   density                5318 non-null   float64
 8   pH                     5318 non-null   float64
 9   sulphates              5318 non-null   float64
 10  alcohol                5318 non-null   float64
 11  quality                5318 non-null   int64  
 12  quality_label          5318 non-null   object 
 13  encoded_quality_label  5318 non-null   int64  
dtypes: float64(11), int64(2), object(1)
memory usage: 581.8+

In [47]:
#combined_wine["encoded_quality_label"]=combined_wine.encoded_quality_label.astype("category")

In [13]:
#combined_wine["quality_label"]=combined_wine.quality_label.astype("category")

In [55]:
from sklearn.model_selection import train_test_split

Split the data into training and testing sets

In [57]:
from sklearn.model_selection import train_test_split, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier

X = combined.drop(["encoded_quality_label", "quality_label", "quality"], axis=1)
y = combined["encoded_quality_label"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# StratifiedKFold with 4 folds
stratified_kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=42)


Explore the set

In [58]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(4254, 11)
(4254,)
(1064, 11)
(1064,)


In [130]:
print("Shape of wines:", X_train.shape)
print("Shape of quality:", y_train.shape)

Shape of wines: (4254, 11)
Shape of quality: (4254,)


In [131]:
X_test.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4697            6.6              0.23         0.32             1.5      0.041   
5032            6.4              0.30         0.27             4.4      0.055   
3750            6.1              0.16         0.37             1.1      0.031   
1634            7.5              0.27         0.31             5.8      0.057   
555             8.9              0.29         0.35             1.9      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
4697                  8.0                  72.0  0.98949  3.22       0.39   
5032                 17.0                 135.0  0.99250  3.23       0.44   
3750                 37.0                  97.0  0.99220  3.40       0.58   
1634                131.0                 313.0  0.99460  3.18       0.59   
555                  25.0                  57.0  0.99700  3.18       1.36   

      alcohol  
4697     12.7  
5032     12.2  
3750     10.5  
1634     10.5  
555      10.3

In [132]:
y_train.head()

3842    1
511     1
62      0
4756    0
3683    0
Name: encoded_quality_label, dtype: int64

My data is skewed--> feature scaling-->normalisation technique:MinMaxScaler


Fits scaler on training data,transforms training and testing data

In [ ]:

from sklearn.preprocessing import MinMaxScaler

norm = MinMaxScaler().fit(X_train)
X_train_norm = norm.transform(X_train)
X_test_norm = norm.transform(X_test)

Train the model & Evaluate the model (Create and fit the RandomForestClassifier with class weights using cross-validation)

In [66]:
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import cohen_kappa_score

rf = RandomForestClassifier(n_estimators=50 , min_samples_split= 10, min_samples_leaf= 4 , max_depth= 40)
cross_val_scores = cross_val_score(rf, X_train, y_train, cv=stratified_kfold, scoring='accuracy')

# Prints the accuracy FOR EACH fold
for fold, accuracy in enumerate(cross_val_scores, 1):
    print(f"Accuracy on Fold {fold}: {accuracy * 100}")

# AVERAGE accuracy across folds
average_accuracy = cross_val_scores.mean()
print("Average Accuracy with class weights and StratifiedKFold:", average_accuracy * 100)

# FitS the model on the entire training set
rf.fit(X_train, y_train)

# Predictions on the test set
y_pred = rf.predict(X_test)

# CalculateS and printS the accuracy on the TEST set
test_accuracy = accuracy_score(y_test, y_pred)
print("Accuracy on the test set:", test_accuracy * 100) 
print("Training Accuracy:", rf.score(X_train, y_train))
#print("Cohens kappa:",rf, X_test, y_test)

Accuracy on Fold 1: 73.7781954887218
Accuracy on Fold 2: 70.86466165413535
Accuracy on Fold 3: 73.47130761994356
Accuracy on Fold 4: 74.03574788334902
Average Accuracy with class weights and StratifiedKFold: 73.03747816153744
Accuracy on the test set: 75.09398496240601
Training Accuracy: 0.8982134461683122


A balanced accuracy of around 50% indicates that the model is performing no better than random chance on your dataset. In the context of classification tasks, 50% balanced accuracy suggests that the model is not effectively discriminating between classes. 

In [163]:
from sklearn.metrics import balanced_accuracy_score
balanced_acc = balanced_accuracy_score(y_test, y_pred)
print("Balanced Accuracy with class weights:", balanced_acc * 100)

Balanced Accuracy with class weights: 49.61771339266497


Confusion Matrix for  Random Forest

In [133]:
from sklearn.metrics import confusion_matrix

y_pred = rf.predict(X_test)
conf_matrix = confusion_matrix(y_test, y_pred)

print("Confusion Matrix Random Forest:")
print(conf_matrix)

Confusion Matrix Random Forest:
[[255 128   0]
 [108 543   0]
 [  0  30   0]]


Random Forest:
True Positives (TP):

Class 1 (low quality): 255 instances correctly predicted.
Class 2 (medium quality): 543 instances correctly predicted.
Class 3 (high quality): 0 instances correctly predicted.
30 false negatives --> predicted as medium quality but actually high 

In [ ]:
from sklearn.metrics import f1_score

y_pred = rf.predict(X_test)
f1 = f1_score(y_test, y_pred, average='weighted')  

print("Weighted F1 Score for RandomForest:", f1*100)


Weighted F1 Score for RandomForest: 73.75517858827631


A weighted F1 score of 0.7375 indicates the overall performance of your Random Forest model in terms of precision and recall, considering the class imbalance in your dataset. The weighted F1 score takes into account the contribution of each class based on its proportion in the dataset.

In [ ]:
from sklearn.metrics import cohen_kappa_score

y_pred = rf.predict(X_test)
kappa = cohen_kappa_score(y_test, y_pred)

print("Cohen's Kappa for RandomForest:", kappa)


Cohen's Kappa for RandomForest: 0.47267456159309584


Hyperparameter tuning for RandomForest ( RandomizedSearchCV)

In [71]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn.ensemble import RandomForestClassifier

# parameter grid
param_dist = {
    'n_estimators': [50, 100, 200, 300, 400],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

rf = RandomForestClassifier(random_state=42)
random_search = RandomizedSearchCV(
    rf,
    param_distributions=param_dist,
    n_iter=10,  
    scoring='accuracy',  
    cv=4,  # 4 cross-validation folds
    random_state=42
)

# Fits RandomizedSearchCV to the training data
random_search.fit(X_train, y_train)

# Gets the best hyperparameters
best_params = random_search.best_params_

# Trains a model with the best hyperparameters
best_rf_model = RandomForestClassifier(random_state=42, **best_params)
best_rf_model.fit(X_train, y_train)

# Evaluates the model on the test set
y_pred = best_rf_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_pred)

print("Best Hyperparameters:", best_params)
print("Accuracy on Test Set:", test_accuracy)


KeyboardInterrupt: 

A Cohen's Kappa coefficient of 1.0 indicates perfect agreement between the predicted labels and the true labels in your Random Forest model.

In [74]:
# Creates a function that shows the Accuracy and the Cross-Validation Accuracy of different Machine Learning Models 

def cross_validate_models(models, X_train, y_train):
    for model in models:
        print(f'{model.__class__.__name__}:')

        # Perform k-fold cross-validation
        stratified_kfold = StratifiedKFold(n_splits=4, shuffle=True, random_state=42)
        cross_val_acc = cross_val_score(model, X_train, y_train, cv=stratified_kfold, scoring='accuracy')

        print(f'Cross-Validation Accuracy: {cross_val_acc.mean() * 100:.2f}%')       
def models(X_train,y_train, X_test, y_test):
#Using KNeighborsClassifier Method of neighbors class to use Nearest Neighbor algorithm
  from sklearn.neighbors import KNeighborsClassifier
  knn= KNeighborsClassifier(n_neighbors = 5, weights = 'uniform', metric = 'minkowski', p = 1)
  knn.fit(X_train, y_train)  

  # Using SVC method of svm class to use Support Vector Machine Algorithm
  from sklearn.svm import SVC
  svc_lin = SVC(kernel = 'linear', random_state = 0, class_weight="balanced")
  svc_lin.fit(X_train, y_train)

  # #Using SVC method of svm class to use Kernel SVM Algorithm
  from sklearn.svm import SVC
  svc_rbf = SVC(kernel = 'rbf', random_state = 0)
  svc_rbf.fit(X_train, y_train)

  # #Using GaussianNB method of naïve_bayes class to use Naïve Bayes Algorithm
  from sklearn.naive_bayes import GaussianNB
  gauss = GaussianNB()
  gauss.fit(X_train, y_train)

  print('K Nearest Neighbor Training Accuracy:', knn.score(X_train, y_train))
  print('Support Vector Machine (Linear Classifier) Training Accuracy:', svc_lin.score(X_train, y_train))
  print('Support Vector Machine (RBF Classifier) Training Accuracy:', svc_rbf.score(X_train, y_train))
  print('Gaussian Naive Bayes Training Accuracy:', gauss.score(X_train, y_train))

  return knn, svc_lin, svc_rbf, gauss 

knn_model, svc_lin_model, svc_rbf_model, gauss_model = models(X_train, y_train, X_test, y_test)
cross_validate_models([knn_model, svc_lin_model, svc_rbf_model, gauss_model], X_train, y_train)

K Nearest Neighbor Training Accuracy: 0.7569346497414199
Support Vector Machine (Linear Classifier) Training Accuracy: 0.5397273154677951
Support Vector Machine (RBF Classifier) Training Accuracy: 0.5961448048895157
Gaussian Naive Bayes Training Accuracy: 0.6391631405735778
KNeighborsClassifier:
Cross-Validation Accuracy: 63.33%
SVC:
Cross-Validation Accuracy: 53.48%
SVC:
Cross-Validation Accuracy: 59.31%
GaussianNB:
Cross-Validation Accuracy: 63.47%


Balanced Accuracy 

In [75]:
from sklearn.metrics import balanced_accuracy_score

def evaluate_models(models, X_train, y_train, X_test, y_test):
    for model in models:
        model_name = model.__class__.__name__

        # Train the model
        model.fit(X_train, y_train)

        # Predictions on the test set
        y_pred = model.predict(X_test)

        # Calculate and print balanced accuracy
        balanced_acc = balanced_accuracy_score(y_test, y_pred)
        print(f'Balanced Accuracy for {model_name}: {balanced_acc * 100:.2f}%')

models_list = [knn_model, svc_lin_model, svc_rbf_model, gauss_model]
evaluate_models(models_list, X_train, y_train, X_test, y_test)


Balanced Accuracy for KNeighborsClassifier: 41.88%
Balanced Accuracy for SVC: 63.49%
Balanced Accuracy for SVC: 34.60%
Balanced Accuracy for GaussianNB: 49.97%


Compare Training and Test Accuracies knn:

In [76]:
from sklearn.metrics import accuracy_score


y_train_pred = knn_model.predict(X_train)
train_accuracy = accuracy_score(y_train, y_train_pred)
print("Training Accuracy:", train_accuracy)


y_test_pred = knn_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_test_pred)
print("Test Accuracy:", test_accuracy)


if train_accuracy > test_accuracy:
    print("Potential overfitting: Training accuracy is higher than test accuracy.")
else:
    print("Balanced performance: Training and test accuracies are comparable.")

Training Accuracy: 0.7569346497414199
Test Accuracy: 0.6456766917293233
Potential overfitting: Training accuracy is higher than test accuracy.


Hyperparamether tuning for knn model

In [ ]:
from sklearn.model_selection import RandomizedSearchCV, train_test_split
import numpy as np

param_dist = {
    'n_neighbors': np.random.randint(1, 20, 10),  # Random integers from 1 to 20
    'weights': ['uniform', 'distance'],
    'p': [1, 2]  # 1 for Manhattan distance, 2 for Euclidean distance
}

# Step 4: Perform RandomizedSearchCV with cross-validation
random_search = RandomizedSearchCV(knn_model, param_distributions=param_dist, n_iter=10, cv=5, scoring='accuracy', random_state=42)
random_search.fit(X_train, y_train)

# Step 5: Get the best hyperparameters
best_params = random_search.best_params_
print("Best Hyperparameters for knn:", best_params)

# Step 6: Evaluate the model on the test set
best_knn_model = random_search.best_estimator_
test_accuracy = best_knn_model.score(X_test, y_test)
print("Test Accuracy with Best Model:", test_accuracy)


Best Hyperparameters for knn: {'weights': 'distance', 'p': 1, 'n_neighbors': 13}
Test Accuracy with Best Model: 0.6644736842105263


Confusion matriy for knn

In [ ]:
y_pred_1 = knn_model.predict(X_test)
conf_matrix_1 = confusion_matrix(y_test, y_pred_1)

print("Confusion Matrix knn:")
print(conf_matrix_1)

Confusion Matrix knn:
[[187 196   0]
 [151 500   0]
 [  4  26   0]]


151 medium quality wines predictes as low

In [ ]:

y_pred_1 = knn_model.predict(X_test)
kappa = cohen_kappa_score(y_test, y_pred)

print("Cohen's Kappa for KNN:", kappa)

Cohen's Kappa for KNN: 0.47267456159309584


Compare Training and Test Accuracies svc; WITHOUT CROSS-VALIDATION:

In [ ]:
svc_lin_model

y_train_pred = svc_lin_model.predict(X_train)
train_accuracy = accuracy_score(y_train, y_train_pred)
print("Training Accuracy:", train_accuracy)


y_test_pred = svc_lin_model.predict(X_test)
test_accuracy = accuracy_score(y_test, y_test_pred)
print("Test Accuracy:", test_accuracy)


#if train_accuracy > test_accuracy:
    print("Potential overfitting: Training accuracy is higher than test accuracy.")
#else:
    print("Balanced performance: Training and test accuracies are comparable.")

Training Accuracy: 0.7226140103432064
Test Accuracy: 0.7124060150375939
Potential overfitting: Training accuracy is higher than test accuracy.


Confusion Matrix for svc 

In [ ]:
y_pred2 = svc_lin_model.predict(X_test)
conf_matrix2 = confusion_matrix(y_test, y_pred2)

print("Confusion Matrix svc:")
print(conf_matrix2)

Confusion Matrix svc:
[[225 158   0]
 [118 533   0]
 [  0  30   0]]


In [ ]:

y_pred2 = svc_lin_model.predict(X_test)
f1 = f1_score(y_test, y_pred2, average='weighted')  

print("Weighted F1 Score for svc:", f1)


Weighted F1 Score for svc: 0.6984988878419696


In [ ]:

y_pred2 = svc_lin_model.predict(X_test)
kappa = cohen_kappa_score(y_test, y_pred2)

print("Cohen's Kappa for svc:", kappa)

Cohen's Kappa for svc: 0.38725825999894603


Confusion Matrix with Cross VAl

In [154]:
from sklearn.metrics import confusion_matrix
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.ensemble import RandomForestClassifier

def print_confusion_matrix(model, X_test, y_test):
    y_pred = model.predict(X_test)


    cm = confusion_matrix(y_test, y_pred)
    print(f'Confusion Matrix for {model.__class__.__name__}:')
    print(cm)
    print("\n")


print_confusion_matrix(knn_model, X_test, y_test)
print_confusion_matrix(svc_lin_model, X_test, y_test)
print_confusion_matrix(svc_rbf_model, X_test, y_test)
print_confusion_matrix(gauss_model, X_test, y_test)




Confusion Matrix for KNeighborsClassifier:
[[187 196   0]
 [151 500   0]
 [  4  26   0]]


Confusion Matrix for SVC:
[[287  76  20]
 [183 253 215]
 [  0   7  23]]


Confusion Matrix for SVC:
[[ 24 359   0]
 [ 16 635   0]
 [  0  30   0]]


Confusion Matrix for GaussianNB:
[[232 149   2]
 [147 473  31]
 [  0  25   5]]




F1 Score with Cross-Validation:

In [155]:
from sklearn.metrics import f1_score

def print_f1_score(model, X_test, y_test):
   
    y_pred = model.predict(X_test)

  
    f1 = f1_score(y_test, y_pred, average='weighted')  

    
    print(f'F1 Score for {model.__class__.__name__}: {f1:.4f}\n')

print_f1_score(knn_model, X_test, y_test)
print_f1_score(svc_lin_model, X_test, y_test)
print_f1_score(svc_rbf_model, X_test, y_test)
print_f1_score(gauss_model, X_test, y_test)


F1 Score for KNeighborsClassifier: 0.6313

F1 Score for SVC: 0.5604

F1 Score for SVC: 0.5048

F1 Score for GaussianNB: 0.6693



Cohen´s Cappa Cross Val

In [77]:
from sklearn.metrics import cohen_kappa_score

def print_cohens_kappa(model, X_test, y_test):

    y_pred = model.predict(X_test)
    kappa = cohen_kappa_score(y_test, y_pred)
    print(f"Cohen's Kappa for {model.__class__.__name__}: {kappa:.4f}\n")

print_cohens_kappa(knn_model, X_test, y_test)
print_cohens_kappa(svc_lin_model, X_test, y_test)
print_cohens_kappa(svc_rbf_model, X_test, y_test)
print_cohens_kappa(gauss_model, X_test, y_test)




Cohen's Kappa for KNeighborsClassifier: 0.2447

Cohen's Kappa for SVC: 0.2654

Cohen's Kappa for SVC: 0.0427

Cohen's Kappa for GaussianNB: 0.3329

